app/src/main/java/com/codesoom/application/ProductService.java

```diff
@@ -0,0 +1,58 @@
+package com.codesoom.application;
+
+import com.codesoom.domain.Product;
+import com.codesoom.exception.ProductNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+public class ProductService {
+    private final Map<Long, Product> products = new LinkedHashMap<>();
+    private Long newId = 0L;
+
+    public List<Product> getProducts() {
+        return new ArrayList<Product>(products.values());
+    }
+
+    public Optional<Product> getProduct(Long id) {
+        Optional<Product> product = findProduct(id);
+        if (!product.isPresent()) {
+            throw new ProductNotFoundException("요청한 " + id + "의 Product가 없습니다.");
+        }
```

Comment:
"없습니다"라고 단언하는 것은 때로는 곤란한 일이 될 수 있습니다. DB에 분명히 있는데도, 접속이 끊어져서 없는 줄 알았다던가 버그가 있어서 없는 것 처럼 나온다던가 하는 경우도 있을 수 있죠. 그래서 수많은 프로그램들은 이런 경우에 "없습니다"라고 하는 대신 "찾지 못했습니다" 같은 표현을 씁니다.

```suggestion
        if (!product.isPresent()) {
            throw new ProductNotFoundException("요청한 " + id + "의 Product를 찾지 못했습니다.");
        }
```

Reply:
log를 보고 오류를 보다 더 잘 찾을 수 있도록 친절하게 적어야겠네요!